File: cpp/questions/reverseKGroup.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode dummy(0, head);
        ListNode* fn = &dummy;
        ListNode* en = fn;
        for (int i=0; i<k; ++i) {
            fn = fn->next;
        }
        bool nr = fn ? true : false;
        while (nr) {
            ListNode* ne = en->next;
            ListNode* nf = fn->next;
            ListNode* cur = nullptr;
            ListNode* nxt = en->next;
            en->next = fn;
            for (int i=0; i<k; ++i) {
                ListNode* nnxt = nxt->next;
                nxt->next = cur;
                cur = nxt;
                nxt = nnxt;
                fn = fn ? fn->next : nullptr;
            }
            nr = fn ? true : false;
            en = ne;
            en->next = nf;
        }
        return dummy.next;
    }
};

File: python/questions/reverseKGroup.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)        
        fn = dummy
        en = dummy
        for _ in range(k):
            fn = fn.next
        nr = True if fn else False
        while nr:
            nf = fn.next
            ne = en.next
            cur = None
            nxt = en.next
            en.next = fn
            for _ in range(k):
                nxtn = nxt.next
                nxt.next = cur
                fn = fn.next if fn else None
                cur = nxt
                nxt = nxtn
            nr = True if fn else False
            en = ne
            ne.next = nf
        return dummy.next
